File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/smartcollection/SmartCollectionRule.kt
/*
 * SmartCollectionRule.kt
 *
 * v1 grammar for `smart_collections.rule_json`. AND-of-clauses
 * shape, no OR, no nesting — per brief §3. Six clause types ship
 * in v1; three (handwriting / signature / has-OCR) are placeholders
 * for the Phase E OCR-derived signals.
 *
 * Wire format is JSON, sealed `type` discriminator. The Kotlin
 * model round-trips both ways via kotlinx.serialization with a
 * polymorphic class discriminator. The same JSON is stored in the
 * DB column and synced to Drive as the smart-collection payload's
 * `rule_json` field.
 */

@file:OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)

package app.quickink.mobile.data.smartcollection

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonClassDiscriminator

/** One AND clause inside a smart-collection rule.
 *
 *  The `type` JSON field is the discriminator (e.g. "folder_is",
 *  "tag_is", "date_range"). The polymorphic serializer matches it
 *  against the @SerialName on each subclass. */
@JsonClassDiscriminator("type")
@Serializable
sealed class RuleClause {

    @Serializable
    @SerialName("folder_is")
    data class FolderIs(
        @SerialName("folder_id") val folderId: String,
    ) : RuleClause()

    @Serializable
    @SerialName("tag_is")
    data class TagIs(
        @SerialName("tag_id") val tagId: String,
    ) : RuleClause()

    @Serializable
    @SerialName("tag_is_not")
    data class TagIsNot(
        @SerialName("tag_id") val tagId: String,
    ) : RuleClause()

    @Serializable
    @SerialName("date_range")
    data class DateRange(
        @SerialName("field")  val field:  String, // "created_at" | "last_opened_at"
        @SerialName("preset") val preset: String, // "today" | "yesterday" | "this_week" | "this_month" | "last_30_days" | "this_quarter"
    ) : RuleClause()

    @Serializable
    @SerialName("source_is")
    data class SourceIs(
        @SerialName("value") val value: String, // "scan" | "import" | "photo" | "video" | "share-extension"
    ) : RuleClause()

    /**
     * OCR-derived signals. Out of scope for the C.3 ship (no flag
     * columns exist yet); the clause parses but its evaluator
     * matches nothing. Lights up in Phase E.
     */
    @Serializable
    @SerialName("has_handwriting")
    data class HasHandwriting(
        @SerialName("value") val value: Boolean,
    ) : RuleClause()

    @Serializable
    @SerialName("has_signature")
    data class HasSignature(
        @SerialName("value") val value: Boolean,
    ) : RuleClause()

    @Serializable
    @SerialName("has_ocr_text")
    data class HasOcrText(
        @SerialName("value") val value: Boolean,
    ) : RuleClause()
}

object SmartCollectionRule {

    /** Canonical JSON config — matches the manifest serializer. */
    private val json: Json = Json {
        encodeDefaults = true
        classDiscriminator = "type"
        ignoreUnknownKeys = true
    }

    /** Encode a list of clauses to the JSON string stored on disk. */
    fun encode(clauses: List<RuleClause>): String =
        json.encodeToString(kotlinx.serialization.builtins.ListSerializer(RuleClause.serializer()), clauses)

    /**
     * Decode a stored rule_json blob. Returns an empty list on any
     * parse failure rather than throwing — a malformed rule turns
     * into "matches nothing" instead of crashing the screen.
     */
    fun decode(text: String?): List<RuleClause> {
        if (text.isNullOrBlank()) return emptyList()
        return runCatching {
            json.decodeFromString(
                kotlinx.serialization.builtins.ListSerializer(RuleClause.serializer()),
                text,
            )
        }.getOrElse { emptyList() }
    }
}

/**
 * Flat editor-input projection of a rule-clause list. Each field
 * is one editable slot in the smart-collection editor; converting
 * to / from a [RuleClause] list lives in [toClauses] /
 * [fromClauses] so the UI doesn't have to reason about the JSON
 * grammar directly.
 *
 * The DateRange clause's `field` is collapsed to `"created_at"`
 * here — the editor doesn't yet expose the `last_opened_at`
 * variant, and a UI for picking the field would need a clearer
 * use case than "the rule grammar supports it".
 *
 * `null` / empty values mean "no clause of that kind". Round-trip
 * is lossless for the slots we surface; clauses the editor
 * doesn't know about (e.g. a date range over `last_opened_at`)
 * are dropped on round-trip — same posture as the JSON decoder's
 * `ignoreUnknownKeys = true`.
 */
data class SmartCollectionRuleInput(
    val folderId: String? = null,
    val datePreset: String? = null,
    val tagIncludeIds: List<String> = emptyList(),
    val tagExcludeIds: List<String> = emptyList(),
    val sourceValue: String? = null,
    val hasHandwriting: Boolean? = null,
    val hasSignature: Boolean? = null,
    val hasOcrText: Boolean? = null,
) {
    /** True when no clause is selected — the editor's Save guard. */
    val isEmpty: Boolean
        get() = folderId == null &&
            datePreset == null &&
            tagIncludeIds.isEmpty() &&
            tagExcludeIds.isEmpty() &&
            sourceValue == null &&
            hasHandwriting == null &&
            hasSignature == null &&
            hasOcrText == null

    /** Compile the input back into the canonical AND-of-clauses list. */
    fun toClauses(): List<RuleClause> = buildList {
        folderId?.let { add(RuleClause.FolderIs(it)) }
        datePreset?.let { add(RuleClause.DateRange(field = "created_at", preset = it)) }
        for (id in tagIncludeIds) add(RuleClause.TagIs(id))
        for (id in tagExcludeIds) add(RuleClause.TagIsNot(id))
        sourceValue?.let { add(RuleClause.SourceIs(it)) }
        hasHandwriting?.let { add(RuleClause.HasHandwriting(it)) }
        hasSignature?.let { add(RuleClause.HasSignature(it)) }
        hasOcrText?.let { add(RuleClause.HasOcrText(it)) }
    }

    companion object {
        /**
         * Build a flat editor-input from a decoded clause list. When
         * a clause type appears more than once the editor only
         * tracks the first (folder / date / source / OCR flags) or
         * unions ids (tag include / exclude).
         */
        fun fromClauses(clauses: List<RuleClause>): SmartCollectionRuleInput {
            var folderId: String? = null
            var datePreset: String? = null
            val includes = mutableListOf<String>()
            val excludes = mutableListOf<String>()
            var source: String? = null
            var hand: Boolean? = null
            var sig: Boolean? = null
            var ocr: Boolean? = null
            for (c in clauses) {
                when (c) {
                    is RuleClause.FolderIs       -> if (folderId == null) folderId = c.folderId
                    is RuleClause.DateRange      -> if (datePreset == null && c.field == "created_at") datePreset = c.preset
                    is RuleClause.TagIs          -> if (c.tagId !in includes) includes += c.tagId
                    is RuleClause.TagIsNot       -> if (c.tagId !in excludes) excludes += c.tagId
                    is RuleClause.SourceIs       -> if (source == null) source = c.value
                    is RuleClause.HasHandwriting -> if (hand   == null) hand   = c.value
                    is RuleClause.HasSignature   -> if (sig    == null) sig    = c.value
                    is RuleClause.HasOcrText     -> if (ocr    == null) ocr    = c.value
                }
            }
            return SmartCollectionRuleInput(
                folderId       = folderId,
                datePreset     = datePreset,
                tagIncludeIds  = includes,
                tagExcludeIds  = excludes,
                sourceValue    = source,
                hasHandwriting = hand,
                hasSignature   = sig,
                hasOcrText     = ocr,
            )
        }
    }
}

File: apps/quickink/ios/QuickInk/Data/SmartCollectionRule.swift
/*
 * SmartCollectionRule.swift
 *
 * v1 rule grammar for `smart_collections.rule_json`. AND-of-clauses
 * shape, no OR, no nesting — per brief §3. Six clause types ship
 * in v1; three (handwriting / signature / has-OCR) are placeholders
 * for the Phase E OCR-derived signals.
 *
 * Wire format is JSON with a `"type"` discriminator. JSON round-trips
 * through canonical-JSON the same byte-for-byte as Android.
 *
 * Mirror of `SmartCollectionRule.kt` in QuickInk's Android target.
 */

import Foundation

public enum RuleClause: Codable, Equatable, Sendable {
    case folderIs(folderId: String)
    case tagIs(tagId: String)
    case tagIsNot(tagId: String)
    case dateRange(field: String, preset: String)
    case sourceIs(value: String)
    case hasHandwriting(value: Bool)
    case hasSignature(value: Bool)
    case hasOcrText(value: Bool)

    private enum CodingKeys: String, CodingKey {
        case type
        case folderId  = "folder_id"
        case tagId     = "tag_id"
        case field
        case preset
        case value
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "folder_is":
            self = .folderIs(folderId: try c.decode(String.self, forKey: .folderId))
        case "tag_is":
            self = .tagIs(tagId: try c.decode(String.self, forKey: .tagId))
        case "tag_is_not":
            self = .tagIsNot(tagId: try c.decode(String.self, forKey: .tagId))
        case "date_range":
            self = .dateRange(
                field:  try c.decode(String.self, forKey: .field),
                preset: try c.decode(String.self, forKey: .preset),
            )
        case "source_is":
            self = .sourceIs(value: try c.decode(String.self, forKey: .value))
        case "has_handwriting":
            self = .hasHandwriting(value: try c.decode(Bool.self, forKey: .value))
        case "has_signature":
            self = .hasSignature(value: try c.decode(Bool.self, forKey: .value))
        case "has_ocr_text":
            self = .hasOcrText(value: try c.decode(Bool.self, forKey: .value))
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: c,
                debugDescription: "Unknown rule type: \(type)",
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .folderIs(let id):
            try c.encode("folder_is", forKey: .type)
            try c.encode(id, forKey: .folderId)
        case .tagIs(let id):
            try c.encode("tag_is", forKey: .type)
            try c.encode(id, forKey: .tagId)
        case .tagIsNot(let id):
            try c.encode("tag_is_not", forKey: .type)
            try c.encode(id, forKey: .tagId)
        case .dateRange(let f, let p):
            try c.encode("date_range", forKey: .type)
            try c.encode(f, forKey: .field)
            try c.encode(p, forKey: .preset)
        case .sourceIs(let v):
            try c.encode("source_is", forKey: .type)
            try c.encode(v, forKey: .value)
        case .hasHandwriting(let v):
            try c.encode("has_handwriting", forKey: .type)
            try c.encode(v, forKey: .value)
        case .hasSignature(let v):
            try c.encode("has_signature", forKey: .type)
            try c.encode(v, forKey: .value)
        case .hasOcrText(let v):
            try c.encode("has_ocr_text", forKey: .type)
            try c.encode(v, forKey: .value)
        }
    }
}

public enum SmartCollectionRule {
    public static func encode(_ clauses: [RuleClause]) -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(clauses),
              let s = String(data: data, encoding: .utf8) else {
            return "[]"
        }
        return s
    }

    public static func decode(_ json: String?) -> [RuleClause] {
        guard let json, !json.isEmpty,
              let data = json.data(using: .utf8) else {
            return []
        }
        return (try? JSONDecoder().decode([RuleClause].self, from: data)) ?? []
    }
}

/// Flat editor-input projection of a rule-clause list. Each field
/// is one editable slot in the smart-collection editor; converting
/// to / from a `[RuleClause]` lives here so the UI doesn't have to
/// reason about the JSON grammar directly. Mirror of Android's
/// `SmartCollectionRuleInput`.
///
/// The `dateRange` clause's `field` is collapsed to `"created_at"`
/// here — the editor doesn't yet expose the `last_opened_at`
/// variant. `nil` / empty values mean "no clause of that kind".
public struct SmartCollectionRuleInput: Equatable, Sendable {
    public var folderId: String?
    public var datePreset: String?
    public var tagIncludeIds: [String]
    public var tagExcludeIds: [String]
    public var sourceValue: String?
    public var hasHandwriting: Bool?
    public var hasSignature: Bool?
    public var hasOcrText: Bool?

    public init(
        folderId: String? = nil,
        datePreset: String? = nil,
        tagIncludeIds: [String] = [],
        tagExcludeIds: [String] = [],
        sourceValue: String? = nil,
        hasHandwriting: Bool? = nil,
        hasSignature: Bool? = nil,
        hasOcrText: Bool? = nil
    ) {
        self.folderId = folderId
        self.datePreset = datePreset
        self.tagIncludeIds = tagIncludeIds
        self.tagExcludeIds = tagExcludeIds
        self.sourceValue = sourceValue
        self.hasHandwriting = hasHandwriting
        self.hasSignature = hasSignature
        self.hasOcrText = hasOcrText
    }

    /// True when no clause is selected — the editor's Save guard.
    public var isEmpty: Bool {
        folderId == nil &&
        datePreset == nil &&
        tagIncludeIds.isEmpty &&
        tagExcludeIds.isEmpty &&
        sourceValue == nil &&
        hasHandwriting == nil &&
        hasSignature == nil &&
        hasOcrText == nil
    }

    /// Compile back into the canonical AND-of-clauses list.
    public func toClauses() -> [RuleClause] {
        var clauses: [RuleClause] = []
        if let folderId { clauses.append(.folderIs(folderId: folderId)) }
        if let datePreset {
            clauses.append(.dateRange(field: "created_at", preset: datePreset))
        }
        for id in tagIncludeIds { clauses.append(.tagIs(tagId: id)) }
        for id in tagExcludeIds { clauses.append(.tagIsNot(tagId: id)) }
        if let sourceValue { clauses.append(.sourceIs(value: sourceValue)) }
        if let hasHandwriting { clauses.append(.hasHandwriting(value: hasHandwriting)) }
        if let hasSignature   { clauses.append(.hasSignature(value: hasSignature)) }
        if let hasOcrText     { clauses.append(.hasOcrText(value: hasOcrText)) }
        return clauses
    }

    /// Build a flat editor-input from a decoded clause list. When
    /// a clause type appears more than once the editor only tracks
    /// the first (folder / date / source / OCR flags) or unions
    /// ids (tag include / exclude).
    public static func fromClauses(_ clauses: [RuleClause]) -> SmartCollectionRuleInput {
        var input = SmartCollectionRuleInput()
        for c in clauses {
            switch c {
            case .folderIs(let id):
                if input.folderId == nil { input.folderId = id }
            case .dateRange(let field, let preset):
                if input.datePreset == nil && field == "created_at" {
                    input.datePreset = preset
                }
            case .tagIs(let id):
                if !input.tagIncludeIds.contains(id) { input.tagIncludeIds.append(id) }
            case .tagIsNot(let id):
                if !input.tagExcludeIds.contains(id) { input.tagExcludeIds.append(id) }
            case .sourceIs(let v):
                if input.sourceValue == nil { input.sourceValue = v }
            case .hasHandwriting(let v):
                if input.hasHandwriting == nil { input.hasHandwriting = v }
            case .hasSignature(let v):
                if input.hasSignature == nil { input.hasSignature = v }
            case .hasOcrText(let v):
                if input.hasOcrText == nil { input.hasOcrText = v }
            }
        }
        return input
    }
}
